### PR TITLE
[UnifiedPDF] Implement "copy" editing command

### PR DESCRIPTION
**Pull request by aprotyas**
#### 1941b5487ff03a1956d7ba96a7fc50753b781b8f
<pre>
[UnifiedPDF] Implement &quot;copy&quot; editing command
<a href="https://bugs.webkit.org/show_bug.cgi?id=269179">https://bugs.webkit.org/show_bug.cgi?id=269179</a>
<a href="https://rdar.apple.com/122763171">rdar://122763171</a>

Reviewed by Sammy Gill.

This patch introduces support for the &quot;copy&quot; editing command. This
command instructs the plugin to copy its current selection into the
general pasteboard. The copy command extracts the plain, HTML, and Web
Archive representations of the current selection, and writes those items
to the NSPasteboardNameGeneral pasteboard.

Currently, we only implement functionality for macOS, though there is
not much that cannot be platform-agnostic in
PDFPluginBase::writeItemsToPasteboard. That will be resolved as part of
<a href="https://webkit.org/b/269174.">https://webkit.org/b/269174.</a>

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.mm:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::writeItemsToPasteboard const):
(WebKit::PDFPluginBase::writeItemsToPasteboard): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::performContextMenuAction):
(WebKit::UnifiedPDFPlugin::isEditingCommandEnabled):
(WebKit::UnifiedPDFPlugin::performCopyEditingOperation const):

Canonical link: <a href="https://commits.webkit.org/274483@main">https://commits.webkit.org/274483@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/206ed12f636616a2e9720190e725852bd0c7cb5b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39236 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18215 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41589 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41770 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15544 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39810 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15333 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13319 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43048 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35628 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35281 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39101 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [loading 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14048 "Build is in progress. Recent messages:") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/11598 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37338 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15654 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8777 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/15317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm [hunk 1]
```diff
@@ -1506,6 +1510,31 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
...
 }
 
+bool UnifiedPDFPlugin::performCopyEditingOperation() const
+{
+#if PLATFORM(MAC)
+    if (!m_currentSelection)
+        return false;
+
+    if (![m_pdfDocument allowsCopying]) {
+        [[NSNotificationCenter defaultCenter] postNotificationName:get_PDFKit_PDFViewCopyPermissionNotification() object:nil];
```
Comment (sammygill) [inline Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:1520]: I created a `WKPDFFormMutationObserver` earlier to use with NSNotificationCenter messages. Maybe we could think about reusing it with a more general name here or a future addition if we don't think it's super important right now

**Review by sammygill — APPROVED** [review]
The link in the commit message looks like it's malformed so probably need to edit the message!

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm (hunk 1 above)
Comment (aprotyas) [reply Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:1520]: It looks like `WKPDFFormMutationObserver` is used to listen to messages. Is there value in reusing it when we want to send a notification? How would we reuse it?

Comment (aprotyas) [reply Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:1520]: Merging for now, but let me know what you think!

**Comment by aprotyas** [comment]
Thanks for the review!

**Pull request merged**